### DISCUSSION TITLE
Agent can't connect to Database

### DISCUSSION THREAD

**Discussion by dgcortespr** [Ylianst/MeshCentral#5456] [General]
**Describe your issue**
After installing the agent on a Windows 11 PC, when I press the connect button I get this message:

**Screenshots**
![image](https://github.com/Ylianst/MeshCentral/assets/148469421/89fa3edb-131f-4c36-a95a-3d6741f49c86)

![image](https://github.com/Ylianst/MeshCentral/assets/148469421/9b3b22b4-a770-4e81-8367-e032d68d603b)

**Server Software (please complete the following information):**
 - OS: Ubuntu 20.04 on hyper-v 
 - Virtualization: 
 - Network: LAN
 - Version: [e.g. 1.0.43]
 - Node: v16

**Client Device (please complete the following information):**
 - Device: Desktop
 - OS: Windows 11
 - Network: Local

{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
    "WANonly": true,
    "_Port": 443,
    "_RedirPort": 80,
    "_AllowLoginToken": true,
    "_AllowFraming": true,
    "_WebRTC": false,
    "_ClickOnce": false,
    "_UserAllowedIP" : "127.0.0.1,::1,192.168.0.100"
  },
  "domains": {
    "": {
      "_title": "MyServer",
      "_title2": "Servername",
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true
    }
  },
  "_letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": "myemail@mydomain.com",
    "names": "myserver.mydomain.com",
    "skipChallengeVerification": true,
    "production": false
  }
}

**Comment by si458** [comment]
this is because you have installed the app already, so its trying to start the app when the database is already opened elsewhere by the service

im guessing you clicked 'Install' then it vanished, then u re-ran the agent.exe and clicked connect?
this is because the 'install' button installs the app as a service in the background so you dont need to click connect,
its one or the other

**Comment by dgcortespr** [comment]
Ok. I uninstall the agent and reinstall it without clicking the connect button. How long does it take for an agent to appear on the web console?

What is the right way to install the agent. 

![image](https://github.com/Ylianst/MeshCentral/assets/148469421/a833bd36-411c-4c46-80c4-61a409cafec5)

**Comment by si458** [comment]
you have wanonly set to true but the serverurl in the screenshot is showing local so the meshagent exe isnt setup correctly,
if the meshcentral is running in ur local network and u only want local devices to connect to it, uncomment wanonly and use lanonly
but if u want remote device external to your network to connect to it, you need to remove wanonly and let it run in wan+lan mode,
then try uninstalling the meshagent from ur remote pc, then redownload the exe from ur webpanel (after u have changed ur settings and restarted meshcentral) then try again

**Reply by dgcortespr** [reply]
Change the setting to use LanOnly for now, reinstall the agent on the computer and it doesn't appear on the web console. 
![image](https://github.com/Ylianst/MeshCentral/assets/148469421/c06b6833-71d9-4b5c-9490-61a6e1fcce24)

{  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
    "_WANonly": true,
    "LANOnly": true,
    "_Port": 443,
    "_RedirPort": 80,
    "_AllowLoginToken": true,
    "_AllowFraming": true,
    "_WebRTC": false,
    "_ClickOnce": false,
    "_UserAllowedIP" : "127.0.0.1,::1,192.168.0.100"
  },
  "domains": {
    "": {
      "title": "UPRB",
      "title2": "RemoteDesktop server",
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true
    }
  },
  "_letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": "myemail@mydomain.com",
    "names": "myserver.mydomain.com",
    "skipChallengeVerification": true,
    "production": false
  }
}

**Comment by dgcortespr** [comment]
Sorry I not used JSON before, what is the instruction I need to write for wan+lan?

**Comment by si458** [comment]
well it depends how you installed it?
docs are here to help you - https://ylianst.github.io/MeshCentral/

**Comment by rmusz** [comment]
as i presume your server link isnt "local" 
I think you have mistakes in config file thats why your installer have also bad config

**Comment by rux2** [comment]
i have the same problem, solutions?
this is my config.json
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
    "_cert": "myserver.mydomain.com",
    "WANonly": true,
    "LANonly": true,
    "_sessionKey": "MyReallySecretPassword1",
    "_port": 443,
    "_aliasPort": 443,
    "_redirPort": 80,
    "_redirAliasPort": 80
  },
  "domains": {
    "": {
      "_title": "MyServer",
      "_title2": "Servername",
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true
    }
  },
  "_letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": "myemail@mydomain.com",
    "names": "myserver.mydomain.com",
    "skipChallengeVerification": true,
    "production": false
  }
}
everytime i go to install an agent on other device, i've the same problem. I've tried to run with windows cmd and return "unable to open database". I've unistalled and reinstalled the aget, but give always the same problem.

**Reply by si458** [reply]
You shouldn't be running anything from the command line?
You simple download the agent exe, double click then click install
You shouldn't be clicking the connect button if you want it to run in the background!
If you see an uninstall button then you already have an agent installed on the computer, so you need to uninstall it from wherever it is installed 

Best way I've found is to open task manager, goto details tab, then look for meahagent.exe unless you have it renamed, then right click and click open file location, it will tell you where it is installed and also to be able to open the exe from that location and click uninstall, then check task manager again to make sure it's really not running

**Reply by rux2** [reply]
thanks for replying, i've done like you said, but unfortunately doesn't work. The process is running in background, but i can't see it in meshcentral dashboard inside of devices.
![Screenshot 2023-10-25 230030](https://github.com/Ylianst/MeshCentral/assets/60732036/f4c1aece-2a8f-463c-b03c-87e7026804f8)
![Screenshot 2023-10-25 230144](https://github.com/Ylianst/MeshCentral/assets/60732036/8e00f401-decd-493e-8b3c-fd3f932903a0)

**Reply by si458** [reply]
Jusy realised you have wanonly AND lanonly set as true in ur config.json, it's one or other or neither.

If all ur machines are inside the same network as the meshcentral server set lanonly to true and wanonly to false

If all ur machines are outside the network ur meshcentral server is in, set wanonly to true and lanonly to false

If u are using a mix of both internal and external machines, just remove both lanonly and wanonly completely from ur config.json!

Note. You might have to uninstall the meshagent from devices and redownload the exe from the webui for devices when u change the settings above

Hope this helps!!!

**Reply by rux2** [reply]
ok, i removed wanonly and lanonly from config.json and now agent status changed from not running to running, but URL server is always local and not meshcentral server ip. Anyway, the device on which i installed the agent does not appear on group list.
![Screenshot 2023-10-26 014747](https://github.com/Ylianst/MeshCentral/assets/60732036/22a9e6bd-dba7-4309-9d6b-22866ef8f55b)
![Screenshot 2023-10-26 014909](https://github.com/Ylianst/MeshCentral/assets/60732036/0afd4709-10cb-4eec-a7d1-01db0e2c112f)

**Reply by si458** [reply]
Right so it's now set in lanonly mode only, so remote devices won't work

You can check this when you run the meshcentral server, it will show you what mode it's running in and its dns/ip name on startup

Are all ur devices going to be local in same network or are you going to have some remotely?

you might need to change _cert to cert and set it as the same name as what's in ur letsencrypt, then check ur remote devices can access the login page of ur server via the dns name you set!

**Reply by rux2** [reply]
I'm new to this world, so i apologize for my stupid questions and errors.
I'm going to have both locally and remotely devices. Deleting wanonly and lanonly , the server is running always in LAN mode. 
I can access to login page simply using https://ip:port, i'm trying it on virtualbox, i don't have dns name. I'll repost my config.json after the previous changes
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
    "cert": "cert",
    "_sessionKey": "MyReallySecretPassword1",
    "_port": 443,
    "_aliasPort": 443,
    "_redirPort": 80,
    "_redirAliasPort": 80
  },
  "domains": {
    "": {
      "_title": "MyServer",
      "_title2": "Servername",
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true
    }
  },
  "_letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": "myemail@mydomain.com",
    "names": "myserver.mydomain.com",
    "skipChallengeVerification": true,
    "production": false
  }
}

**Reply by si458** [reply]
So to be able to use it externally outside ur network you need a dns name pointing to ur router and port forwarded, also the `cert` in`settings` needs to be a valid dns name, also you must have hairpin nat setup so the internal machines can see the port forwarding correctly

Then if you want a valid ssl certificate,  you need to make `names` in `letsencrypt` match `cert`, and also you need to uncomment the letsencrypt section, as its currently commented `_letsencrypt` to `letsencrypt`
